keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,415 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The service controller in the Kubernetes cloud controller manager (KCCM)
+configures the load balancer and its corresponding health check (HC) following
+service related events. The configured health check is then used by the load
+balancer as to determine which instances are candidates for traffic load
+balancing. For certain cloud providers (GCP): the KCCM configures the HC to
+target the Kubernetes service proxy for this information. In this KEP we will
+focus on Kube-proxy, since it's the only service proxy under the responsibility
+of the Kubernetes project.
+
+We can define two classes of services for what concerns the HC:
+
+- `externalTrafficPolicy: Cluster` / `eTP:Cluster` (default)
+- `externalTrafficPolicy: Local` / `eTP:Local`
+
+For class `eTP:Cluster` services: Kube-proxy currently returns an answer
+following its `healthz` state (specifically, whether the data-plane programming
+is known to be stale). For `eTP:Local` services: Kube-proxy only reports if the
+service for which the load balancer was created for, has a `Ready` endpoint
+running on the node. This KEP will only focus on the former case.
+
+This KEP proposes two changes:
+
+1. That Kube-proxy provides a mechanism for load balancers to do connection
+  draining for terminating Nodes. This is to be done by Kube-proxy by inspecting
+  a field on the Node object which indicates that the "node is
+  terminating/deleting" and when seen, starts failing the HC. When discussing
+  this scenario, the primary case this applied to was: downscaling by the
+  cluster autoscaler (CA). Given that the CA taints the Node which is to be
+  downscaled and deleted, with the taint: `ToBeDeletedByClusterAutoscaler` it
+  would seem most appropriate to use that here. It is unfortunate to have this
+  taint spread around the code base, but for now: no better indicator has been
+  thought of. The `.spec.unschedulable` field has been discussed as well.
+  Setting that field is usually followed by eviction and Node termination, but
+  it doesn't have as strong of a direct link to the termination/deletion of the
+  Node, as the taint does. Users can for example decide to cordon all nodes
+  during at given moment in time. Using `.spec.unschedulable` as this signal
```

Comment:
We can debate this in more detail later (doesn't need to block KEP, I think), but ONLY handling CA's taint means we don't get any benefit if not using CA.  I just don't think that is sufficient.

As of now we do not have a clear answer to the question we are asking, so we will have to approximate.

Reply:
+1

it would be good to add figuring this out as a GA criteria maybe?

Reply:
OK, we should figure out the field before going to Beta right? We don't want to enable things by default and not be sure that it's right, right?